File: Examples/WindowsDesktop/Components/DashboardControl.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using UnderAutomation.UniversalRobots;
using UnderAutomation.UniversalRobots.Common;
using UnderAutomation.UniversalRobots.Dashboard;

public partial class DashboardControl : UserControl, IUserControl
{
    UR _ur;
    public DashboardControl(UR ur)
    {
        _ur = ur;

        InitializeComponent();

        // fill enum combos
        foreach (var value in Enum.GetValues(typeof(UserRoles))) cbUserRoles.Items.Add(value);
        cbUserRoles.SelectedItem = UserRoles.Operator;
        foreach (var value in Enum.GetValues(typeof(OperationalModes))) cbOperationalMode.Items.Add(value);
        cbOperationalMode.SelectedItem = OperationalModes.Automatic;
    }

    #region IUserControl
    public string Title => "Dashboard (Remote send commands)";

    public bool FeatureEnabled => !string.IsNullOrEmpty(_ur.Dashboard.IP);

    public void PeriodicUpdate()
    {
        pnlMain.Enabled = FeatureEnabled;
    }

    public void OnClose() { }

    public void OnOpen() { }
    #endregion

    // Append command response to the console text area
    private void Log(CommandResponse response, string command, string messageBefore=null, string messageAfter=null)
    {
        txtConsole.AppendText(command);
        txtConsole.AppendText("()\r\n");

        txtConsole.SelectionBullet = true;
        if (response.Succeed)
        {
            txtConsole.SelectionColor = Color.Green;
            txtConsole.AppendText("OK : ");
        }
        else
        {
            txtConsole.SelectionColor = Color.Red;
            txtConsole.AppendText("FAILED : ");
        }

        if(messageBefore != null) txtConsole.AppendText(messageBefore);
        txtConsole.AppendText(response.ToString() ?? "");
        if (messageAfter != null) txtConsole.AppendText(messageAfter);


        txtConsole.SelectionBullet = false;
        txtConsole.SelectionColor = Color.White;
        txtConsole.AppendText("\r\n\r\n> ");
        txtConsole.ScrollToCaret();
    }

    #region Command clicks
    private void btnGetRobotMode_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetRobotMode();
        Log(response, "GetRobotMode");
    }

    private void btnPowerOn_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.PowerOn();
        Log(response, "PowerOn");
    }

    private void btnPowerOff_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.PowerOff();
        Log(response, "PowerOff");
    }

    private void ReaseBrake_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.ReleaseBrake();
        Log(response, "ReleaseBrake");
    }

    private void btnUnlockProtectiveStop_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.UnlockProtectiveStop();
        Log(response, "UnlockProtectiveStop");
    }

    private void btnShutdown_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.Shutdown();
        Log(response, "Shutdown");
    }

    private void btnLoadProgram_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.LoadProgram(txtProgram.Text);
        Log(response, "LoadProgram");

        Config.Current.LoadProgram = txtProgram.Text;
        Config.Save();
    }

    private void btnGetLoadedProgram_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetLoadedProgram();
        Log(response, "GetLoadedProgram");
    }

    private void btnPlay_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.Play();
        Log(response, "Play");
    }

    private void btnStop_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.Stop();
        Log(response, "Stop");
    }

    private void btnPause_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.Pause();
        Log(response, "Pause");
    }

    private void btnProgrammRunning_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.IsProgramRunning();
        Log(response, "IsProgramRunning");
    }

    private void btnGetProgramState_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetProgramState();
        Log(response, "GetProgramState");
    }

    private void btnIsProgramSaved_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.IsProgramSaved();
        Log(response, "IsProgramSaved");
    }

    private void btnShowPopup_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.ShowPopup(txtPopup.Text);
        Log(response, "ShowPopup");

        Config.Current.ShowPopup = txtPopup.Text;
        Config.Save();
    }

    private void btnClosePopup_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.ClosePopup();
        Log(response, "ClosePopup");
    }

    private void btnAddToLog_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.AddToLog(txtLog.Text);
        Log(response, "AddToLog");

        Config.Current.AddToLog = txtLog.Text;
        Config.Save();
    }

    private void btnPolyscopeVersion_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetPolyscopeVersion();
        Log(response, "GetPolyscopeVersion");
    }

    private void btnLoadInstallation_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.LoadInstallation(txtInstallation.Text);
        Log(response, "LoadInstallation");

        Config.Current.LoadInstallation = txtInstallation.Text;
        Config.Save();
    }

    private void btnGetSerialNumber_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetSerialNumber();
        Log(response, "GetSerialNumber");
    }

    private void btnGetRobotModel_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetRobotModel();
        Log(response, "GetRobotModel");
    }

    private void btnGetOperationalMode_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetOperationalMode();
        Log(response, "GetOperationalMode");
    }

    private void btnClearOIperationalMode_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.ClearOperationalMode();
        Log(response, "ClearOperationalMode");
    }

    private void btnSetOperationalMode_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.SetOperationalMode((OperationalModes)cbOperationalMode.SelectedItem);
        Log(response, "SetOperationalMode");
    }

    private void btnSetUserRole_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.SetUserRole((UserRoles)cbUserRoles.SelectedItem);
        Log(response, "SetUserRole");
    }

    private void btnIsInRemoteControl_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.IsInRemoteControl();
        Log(response, "IsInRemoteControl");
    }

    private void btnSafetyStatus_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetSafetyStatus();
        Log(response, "GetSafetyStatus");
    }

    private void btnCloseSafetyPopup_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.CloseSafetyPopup();
        Log(response, "CloseSafetyPopup");
    }

    private void btnRestartSafety_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.RestartSafety();
        Log(response, "RestartSafety");
    }

    private void txtSendCustomCommand_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.SendCustomDashboardCommand(txtCustomCommand.Text);
        Log(response, "SendCustomDashboardCommand");
    }

    private void btnGetVariableValue_Click(object sender, EventArgs e)
    {
        var response = _ur.Dashboard.GetVariable(txtVariableName.Text);
        if(response.Value.Type == GlobalVariableTypes.None)
        {
            Log(response, "GetVariableValue", $"Variable {txtVariableName.Text} is not declared. ");
        }
        else
        {
            Log(response, "GetVariableValue", response.Succeed ? $"{response.Value.Type} {response.Value.Name} = " : "");
        }
    }
    #endregion

}

File: UnderAutomation.UniversalRobots.Showcase.Forms.vb/Components/DashboardControl.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports UnderAutomation.UniversalRobots
Imports UnderAutomation.UniversalRobots.Common
Imports UnderAutomation.UniversalRobots.Dashboard

Public Partial Class DashboardControl
    Inherits UserControl
    Implements IUserControl
    Private _ur As UR
    Public Sub New(ur As UR)
        _ur = ur

        InitializeComponent()

        ' fill enum combos
        For Each value In [Enum].GetValues(GetType(UserRoles))
            cbUserRoles.Items.Add(value)
        Next
        cbUserRoles.SelectedItem = UserRoles.Operator
        For Each value In [Enum].GetValues(GetType(OperationalModes))
            cbOperationalMode.Items.Add(value)
        Next
        cbOperationalMode.SelectedItem = OperationalModes.Automatic
    End Sub

#Region "IUserControl"
    Public ReadOnly Property Title As String Implements IUserControl.Title
        Get
            Return "Dashboard (Remote send commands)"
        End Get
    End Property

    Public ReadOnly Property FeatureEnabled As Boolean Implements IUserControl.FeatureEnabled
        Get
            Return Not String.IsNullOrEmpty(_ur.Dashboard.IP)
        End Get
    End Property

    Public Sub PeriodicUpdate() Implements IUserControl.PeriodicUpdate
        pnlMain.Enabled = FeatureEnabled
    End Sub

    Public Sub OnClose() Implements IUserControl.OnClose
    End Sub

    Public Sub OnOpen() Implements IUserControl.OnOpen
    End Sub
#End Region

    ' Append command response to the console text area
    Private Sub Log(response As CommandResponse, command As String, Optional messageBefore As String = Nothing, Optional messageAfter As String = Nothing)
        txtConsole.AppendText(command)
        txtConsole.AppendText("()" & Microsoft.VisualBasic.Constants.vbCrLf)

        txtConsole.SelectionBullet = True
        If response.Succeed Then
            txtConsole.SelectionColor = Color.Green
            txtConsole.AppendText("OK : ")
        Else
            txtConsole.SelectionColor = Color.Red
            txtConsole.AppendText("FAILED : ")
        End If

        If Not Equals(messageBefore, Nothing) Then txtConsole.AppendText(messageBefore)
        txtConsole.AppendText(If(response.ToString(), ""))
        If Not Equals(messageAfter, Nothing) Then txtConsole.AppendText(messageAfter)


        txtConsole.SelectionBullet = False
        txtConsole.SelectionColor = Color.White
        txtConsole.AppendText(Microsoft.VisualBasic.Constants.vbCrLf & Microsoft.VisualBasic.Constants.vbCrLf & "> ")
        txtConsole.ScrollToCaret()
    End Sub

#Region "Command clicks"
    Private Sub btnGetRobotMode_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetRobotMode()
        Log(response, "GetRobotMode")
    End Sub

    Private Sub btnPowerOn_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.PowerOn()
        Log(response, "PowerOn")
    End Sub

    Private Sub btnPowerOff_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.PowerOff()
        Log(response, "PowerOff")
    End Sub

    Private Sub ReaseBrake_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.ReleaseBrake()
        Log(response, "ReleaseBrake")
    End Sub

    Private Sub btnUnlockProtectiveStop_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.UnlockProtectiveStop()
        Log(response, "UnlockProtectiveStop")
    End Sub

    Private Sub btnShutdown_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.Shutdown()
        Log(response, "Shutdown")
    End Sub

    Private Sub btnLoadProgram_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.LoadProgram(txtProgram.Text)
        Log(response, "LoadProgram")

        Config.Current.LoadProgram = txtProgram.Text
        Config.Save()
    End Sub

    Private Sub btnGetLoadedProgram_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetLoadedProgram()
        Log(response, "GetLoadedProgram")
    End Sub

    Private Sub btnPlay_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.Play()
        Log(response, "Play")
    End Sub

    Private Sub btnStop_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.Stop()
        Log(response, "Stop")
    End Sub

    Private Sub btnPause_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.Pause()
        Log(response, "Pause")
    End Sub

    Private Sub btnProgrammRunning_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.IsProgramRunning()
        Log(response, "IsProgramRunning")
    End Sub

    Private Sub btnGetProgramState_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetProgramState()
        Log(response, "GetProgramState")
    End Sub

    Private Sub btnIsProgramSaved_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.IsProgramSaved()
        Log(response, "IsProgramSaved")
    End Sub

    Private Sub btnShowPopup_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.ShowPopup(txtPopup.Text)
        Log(response, "ShowPopup")

        Config.Current.ShowPopup = txtPopup.Text
        Config.Save()
    End Sub

    Private Sub btnClosePopup_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.ClosePopup()
        Log(response, "ClosePopup")
    End Sub

    Private Sub btnAddToLog_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.AddToLog(txtLog.Text)
        Log(response, "AddToLog")

        Config.Current.AddToLog = txtLog.Text
        Config.Save()
    End Sub

    Private Sub btnPolyscopeVersion_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetPolyscopeVersion()
        Log(response, "GetPolyscopeVersion")
    End Sub

    Private Sub btnLoadInstallation_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.LoadInstallation(txtInstallation.Text)
        Log(response, "LoadInstallation")

        Config.Current.LoadInstallation = txtInstallation.Text
        Config.Save()
    End Sub

    Private Sub btnGetSerialNumber_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetSerialNumber()
        Log(response, "GetSerialNumber")
    End Sub

    Private Sub btnGetRobotModel_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetRobotModel()
        Log(response, "GetRobotModel")
    End Sub

    Private Sub btnGetOperationalMode_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetOperationalMode()
        Log(response, "GetOperationalMode")
    End Sub

    Private Sub btnClearOIperationalMode_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.ClearOperationalMode()
        Log(response, "ClearOperationalMode")
    End Sub

    Private Sub btnSetOperationalMode_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.SetOperationalMode(CType(cbOperationalMode.SelectedItem, OperationalModes))
        Log(response, "SetOperationalMode")
    End Sub

    Private Sub btnSetUserRole_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.SetUserRole(CType(cbUserRoles.SelectedItem, UserRoles))
        Log(response, "SetUserRole")
    End Sub

    Private Sub btnIsInRemoteControl_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.IsInRemoteControl()
        Log(response, "IsInRemoteControl")
    End Sub

    Private Sub btnSafetyStatus_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetSafetyStatus()
        Log(response, "GetSafetyStatus")
    End Sub

    Private Sub btnCloseSafetyPopup_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.CloseSafetyPopup()
        Log(response, "CloseSafetyPopup")
    End Sub

    Private Sub btnRestartSafety_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.RestartSafety()
        Log(response, "RestartSafety")
    End Sub

    Private Sub txtSendCustomCommand_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.SendCustomDashboardCommand(txtCustomCommand.Text)
        Log(response, "SendCustomDashboardCommand")
    End Sub

    Private Sub btnGetVariableValue_Click(sender As Object, e As EventArgs)
        Dim response = _ur.Dashboard.GetVariable(txtVariableName.Text)
        If response.Value.Type = GlobalVariableTypes.None Then
            Log(response, "GetVariableValue", $"Variable {txtVariableName.Text} is not declared. ")
        Else
            Log(response, "GetVariableValue", If(response.Succeed, $"{response.Value.Type} {response.Value.Name} = ", ""))
        End If
    End Sub
#End Region

End Class
